### ISSUE TITLE
Rhino cannot evaluate a certain JavaScript code

### ISSUE THREAD

**Issue by otsuka-kohei** [mozilla/rhino#1218]
Rhino returns a error `missing ; before statement` at line #20 of [this JavaScript code](https://github.com/receiptline/receiptline/blob/a2137c539d2169569c3efe1fd1be9a6dfd4e3fa3/lib/receiptline.js).
  
Do you know the reason of this error?

**Comment by otsuka-kohei** [comment]
FYI, Chrome and Firefox can execute that JavaScript code.

**Comment by p-bakker** [comment]
Thats a large chunk of code, I think you'll have to put in the effort to pinpoint where the actual issue occurs within the code (line 20 is just a wrapper around all the code, so it happens somewhere inside the wrapper).

You haven't even specified which version of Rhino you're using, so even if someone would be willing to digg into such a broad question, they'd have no clue where to start

**Comment by otsuka-kohei** [comment]
@p-bakker 
Let me apologize for lack of information.
  
I am using Rhino v1.7.14.
  
Also, this line occurs a error.
I found that `let` occurs error, but `const` does not occur error.

```
let foo = "foo";
```

**Comment by otsuka-kohei** [comment]
@p-bakker 
I created a Java project to reproduce the error with minimum code.
This project shows same error.
  
Could you please try to run [this Java project](https://github.com/otsuka-kohei/rhino-user)? 
  
You can run with a command `./gradlew run`.
I built and runned this Java project with Java 11.

**Comment by p-bakker** [comment]
Mmm, strange, can reproduce this, while its quite basic JavaScript stuff

BTW: providing just the code snippet would've been sufficient:
```
	  Context ctx = Context.enter();
      ctx.setOptimizationLevel(-1);
      Scriptable scriptable = ctx.initStandardObjects();
      ctx.evaluateString(scriptable, "let foo = \"foo\";", "<cmd>", 0, null);
```

**Comment by p-bakker** [comment]
Figured it out: you're running Rhino without having specified the language version, so it uses the default (0), in which `let` isn't supported, so it sees 'let' as a identifier declaration and then indeed there's a missing semi-colon.

Setting the language version to 200 fixes it:
```
  Context ctx = Context.enter();
  ctx.setLanguageVersion(200);
  Scriptable scriptable = ctx.initStandardObjects();
  ctx.evaluateString(scriptable, "let foo = \"foo\";\n", "<cmd>", 0, null);
```
 If this also solves the problem for you, please close the case

**Comment by otsuka-kohei** [comment]
@p-bakker
Thank you for investigation.
Rhino can evaluate `let foo = "foo";` by your solution.
  
But, I faced another issue.
Rhino returns error `TypeError: redeclaration of const foo.` when evaluate below JavaScript code.
```javascript
function funcFoo() {
  const foo = "foo";
  let str = foo;
  {
    const foo = "FOO";
    str = str + foo;
  }
  return str;
}
```
`const foo` in code bloack must be different from another `const foo` at head of function.
Of course, Firefox can evaluate this code.

Do you know what is wrong?
Let me apologize for frequent inquiries.

Best Regards,

**Comment by p-bakker** [comment]
Unfortunately Rhino currently doesn't support proper block scoping for `const`, resulting in the redeclaration error.

There are already several issues that report this. Hopefully this issue can be solved in v1.7.15 or v2.0.0.

For now the best path forward is to run your (library) code through Babel, using the 'rhino' profile.

No problem with the questions, but we have more appropriate places for them: discussions, Google Groups and Stack Overflow

**Comment by otsuka-kohei** [comment]
@p-bakker
Thank you for answer.
I understand current situation.
I will consider to use Babel.
I will close this ticket.
Best Regards,
